internal/newstate/graph.go

```diff
@@ -0,0 +1,225 @@
+}
+
+// graph is a graph-like representation of Gateway API resources.
+// It is assumed that the Gateway resource is a single resource.
+type graph struct {
+	// Listeners holds listeners keyed by their names in the Gateway resource.
+	Listeners map[string]*listener
+	// Routes holds route resources.
+	Routes map[types.NamespacedName]*route
+}
+
+// buildGraph builds a graph from a store assuming that the Gateway resource has the gwNsName namespace and name.
+func buildGraph(store *store, gwNsName types.NamespacedName) *graph {
+	listeners := buildListeners(store.gw)
+
+	routes := make(map[types.NamespacedName]*route)
+	for _, ghr := range store.httpRoutes {
+		ignored, r := bindHTTPRouteToListeners(ghr, gwNsName, listeners)
+		if !ignored {
+			routes[getNamespacedName(ghr)] = r
+		}
+	}
+
+	return &graph{
+		Listeners: listeners,
+		Routes:    routes,
+	}
+}
+
+// bindHTTPRouteToListeners tries to bind an HTTPRoute to listener.
+// There are three possibilities:
+// (1) HTTPRoute will be ignored.
+// (2) HTTPRoute will be processed but not bound.
+// (3) HTTPRoute will be processed and bound to a listener.
+func bindHTTPRouteToListeners(
+	ghr *v1alpha2.HTTPRoute,
+	gwNsName types.NamespacedName,
+	listeners map[string]*listener,
+) (ignored bool, r *route) {
+	if len(ghr.Spec.ParentRefs) == 0 {
+		// ignore HTTPRoute without refs
+		return true, nil
+	}
+
+	r = &route{
+		Source:                 ghr,
+		ValidSectionNameRefs:   make(map[string]struct{}),
+		InvalidSectionNameRefs: make(map[string]struct{}),
+	}
+
+	// FIXME (pleshakov) Handle the case when parent refs are duplicated
+
+	ignored = true
+
+	for _, p := range ghr.Spec.ParentRefs {
+		// Step 1 - Ensure the ref references the Gateway and has a non-empty section name
+
+		if ignoreParentRef(p, ghr.Namespace, gwNsName) {
+			continue
+		}
+
+		ignored = false
+
+		name := string(*p.SectionName)
+
+		// Step 2 - Find a listener
+
+		// FIXME(pleshakov)
+		// For now, let's do simple matching.
+		// However, we need to also support wildcard matching.
+		// More over, we need to handle cases when a Route host matches multiple HTTP listeners on the same port when
+		// sectionName is empty and only choose one listener.
+		// For example:
+		// - Route with host foo.example.com;
+		// - listener 1 for port 80 with hostname foo.example.com
+		// - listener 2 for port 80 with hostname *.example.com;
+		// In this case, the Route host foo.example.com should choose listener 2, as it is a more specific match.
+
+		l, exists := listeners[name]
+		if !exists {
+			r.InvalidSectionNameRefs[name] = struct{}{}
+			continue
+		}
+
+		accepted := findAcceptedHostnames(l.Source.Hostname, ghr.Spec.Hostnames)
+
+		if len(accepted) > 0 {
+			for _, h := range accepted {
+				l.AcceptedHostnames[h] = struct{}{}
+			}
+			r.ValidSectionNameRefs[name] = struct{}{}
+			l.Routes[getNamespacedName(ghr)] = r
+		} else {
+			r.InvalidSectionNameRefs[name] = struct{}{}
+		}
+	}
+
+	if ignored {
```

Comment:
Why not just return `ignored, r`?

Reply:
@kate-osborn  I did it so that in case of ignored equal true, r is always nil. Without that check, if all parentRefs are ignored, which makes ignored equal true, the r will not be nil.

Reply:
Ok, but in that case `r` is not used, right?

Reply:
yep, r is not used in that case - the called needs to ignore it, so, it doesn't have to be nil. However, returning nil also makes the test code less verbose - expected value is nil, not a full route, which needs to be defined.

Reply:
I see. Is it necessary to test that the route is nil when ignored is true? Seems unnecessary to me since we are not using the route in that case.

Reply:
the test tests both return values. but it is possible to ignore testing for route if expected ignored is true. but I feel it is a bit more complicated in that case - it easier to test both values. I think it is also better from design perspective - instead of returning an invalid value (which should be ignored), return nil.

Reply:
Ok. I'm not sure I agree that an empty struct is an invalid value, but if you prefer this approach then I'm good with it.